compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1001,31 +1001,35 @@ impl<'a> Parser<'a> {
     ) -> PResult<'a, ItemInfo> {
         let sp_start = self.prev_token.span;
         let abi = self.parse_abi(); // ABI?
-        match self.parse_item_list(attrs, |p| p.parse_foreign_item(ForceCollect::No)) {
-            Ok(items) => {
-                let module = ast::ForeignMod { unsafety, abi, items };
-                Ok((Ident::empty(), ItemKind::ForeignMod(module)))
-            }
-            Err(mut err) => {
-                let current_qual_sp = self.prev_token.span;
-                let current_qual_sp = current_qual_sp.to(sp_start);
-                if let Ok(current_qual) = self.span_to_snippet(current_qual_sp) {
-                    // FIXME(davidtwco): avoid depending on the error message text
-                    if err.message[0].0.expect_str() == "expected `{`, found keyword `unsafe`" {
-                        let invalid_qual_sp = self.token.uninterpolated_span();
-                        let invalid_qual = self.span_to_snippet(invalid_qual_sp).unwrap();
+        if self.token.is_keyword(kw::Unsafe) {
+            let mut err = self.expect(&token::OpenDelim(Delimiter::Brace)).unwrap_err();
+            let current_qual_sp = self.prev_token.span;
+            let current_qual_sp = current_qual_sp.to(sp_start);
+            if let Ok(current_qual) = self.span_to_snippet(current_qual_sp) {
+                let invalid_qual_sp = self.token.uninterpolated_span();
+                let invalid_qual = self.span_to_snippet(invalid_qual_sp).unwrap();
 
-                        err.span_suggestion(
-                                current_qual_sp.to(invalid_qual_sp),
-                                &format!("`{}` must come before `{}`", invalid_qual, current_qual),
-                                format!("{} {}", invalid_qual, current_qual),
-                                Applicability::MachineApplicable,
-                            ).note("keyword order for functions declaration is `pub`, `default`, `const`, `async`, `unsafe`, `extern`");
-                    }
-                }
-                Err(err)
+                err.span_suggestion(
+                        current_qual_sp.to(invalid_qual_sp),
+                        &format!("`{}` must come before `{}`", invalid_qual, current_qual),
+                        format!("{} {}", invalid_qual, current_qual),
+                        Applicability::MachineApplicable,
+                    ).note("keyword order for functions declaration is `pub`, `default`, `const`, `async`, `unsafe`, `extern`");
```

Comment:
"keyword order for functions declaration"

This isn't a function declaration, though, right?

Reply:
That make sense. And I think we can omit this note, since we just want to parse a `foreign mod` here.